test/inductor/test_inductor_freezing.py

```diff
@@ -324,6 +324,45 @@ def foo(mod, x):
                 out_optimized_for_infernece, out_eager, atol=1e-2, rtol=1e-2
             )
 
+    def test_folded_conv_bn_relu6(self):
+        class Model(torch.nn.Module):
```

Comment:
I don't think we're running the conv in lower precision, right now.  

https://github.com/pytorch/pytorch/blob/758948304a34d8f21f0219aba56a59d1d3686de7/torch/_inductor/fx_passes/binary_folding.py#L14-L42

We are only marking `mark_mixed_dtype_conv` if its followed by a series of foldable ops, and then a convert_element_dtype, which doesn't occur in the relu6 example.

We should be unconditionally mark the original dtype in the convs. 

Could you could also add a mechanism so that we only add the convert_element_dtype on weights/bias if we actually fold any of the conv inputs with a higher precision ?

Reply:
@eellison , I double check it, the conv running lower precision after folding:
```
    def forward(self, arg7_1: bf16[3, 3, 32, 32]):
        # No stacktrace found for following nodes
        _frozen_param18: bf16[32] = self._frozen_param18
        _frozen_param19 = self._frozen_param19

        # File: /home/xiaobing/pytorch-offical/test/inductor/test_inductor_freezing.py:101, code: return self.bn(self.conv(x))
        _convolution_pointwise_default: bf16[3, 32, 15, 15] = torch.ops.mkldnn._convolution_pointwise.default(arg7_1, _frozen_param19, _frozen_param18, [0, 0], [2, 2], [1, 1], 1, 'none', [], '');  arg7_1 = _frozen_param19 = _frozen_param18 = None

        # File: /home/xiaobing/pytorch-offical/test/inductor/test_inductor_freezing.py:335, code: return self.relu6(self.conv_bn(x))
        convert_element_type_default: f32[3, 32, 15, 15] = torch.ops.prims.convert_element_type.default(_convolution_pointwise_default, torch.float32);  _convolution_pointwise_default = None
        clamp_min: f32[3, 32, 15, 15] = torch.ops.aten.clamp_min.default(convert_element_type_default, 0.0);  convert_element_type_default = None
        clamp_max: f32[3, 32, 15, 15] = torch.ops.aten.clamp_max.default(clamp_min, 6.0);  clamp_min = None
        convert_element_type_4: bf16[3, 32, 15, 15] = torch.ops.prims.convert_element_type.default(clamp_max, torch.bfloat16);  clamp_max = None

        # No stacktrace found for following nodes
        inductor_force_stride_order_default: bf16[3, 32, 15, 15] = torch.ops.prims.inductor_force_stride_order.default(convert_element_type_4, (7200, 225, 15, 1));  convert_element_type_4 = None
        return (inductor_force_stride_order_default,)
```

see the original  fx graph:

```
2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, arg0_1: bf16[32, 3, 3, 3], arg1_1: bf16[32], arg2_1: bf16[32], arg3_1: bf16[32], arg4_1: bf16[32], arg5_1: bf16[32], arg6_1: i64[], arg7_1: bf16[3, 3, 32, 32]):
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/xiaobing/pytorch-offical/test/inductor/test_inductor_freezing.py:101, code: return self.bn(self.conv(x))
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convolution: bf16[3, 32, 15, 15] = torch.ops.aten.convolution.default(arg7_1, arg0_1, arg1_1, [2, 2], [0, 0], [1, 1], False, [0, 0], 1);  arg7_1 = arg0_1 = arg1_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convert_element_type: f32[32] = torch.ops.prims.convert_element_type.default(arg4_1, torch.float32);  arg4_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convert_element_type_1: f32[32] = torch.ops.prims.convert_element_type.default(arg5_1, torch.float32);  arg5_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         add: f32[32] = torch.ops.aten.add.Tensor(convert_element_type_1, 0.001);  convert_element_type_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         sqrt: f32[32] = torch.ops.aten.sqrt.default(add);  add = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         reciprocal: f32[32] = torch.ops.aten.reciprocal.default(sqrt);  sqrt = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         mul: f32[32] = torch.ops.aten.mul.Tensor(reciprocal, 1);  reciprocal = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze: f32[32, 1] = torch.ops.aten.unsqueeze.default(convert_element_type, -1);  convert_element_type = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_1: f32[32, 1, 1] = torch.ops.aten.unsqueeze.default(unsqueeze, -1);  unsqueeze = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_2: f32[32, 1] = torch.ops.aten.unsqueeze.default(mul, -1);  mul = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_3: f32[32, 1, 1] = torch.ops.aten.unsqueeze.default(unsqueeze_2, -1);  unsqueeze_2 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         sub: f32[3, 32, 15, 15] = torch.ops.aten.sub.Tensor(convolution, unsqueeze_1);  convolution = unsqueeze_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         mul_1: f32[3, 32, 15, 15] = torch.ops.aten.mul.Tensor(sub, unsqueeze_3);  sub = unsqueeze_3 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_4: bf16[32, 1] = torch.ops.aten.unsqueeze.default(arg2_1, -1);  arg2_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_5: bf16[32, 1, 1] = torch.ops.aten.unsqueeze.default(unsqueeze_4, -1);  unsqueeze_4 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         mul_2: f32[3, 32, 15, 15] = torch.ops.aten.mul.Tensor(mul_1, unsqueeze_5);  mul_1 = unsqueeze_5 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_6: bf16[32, 1] = torch.ops.aten.unsqueeze.default(arg3_1, -1);  arg3_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         unsqueeze_7: bf16[32, 1, 1] = torch.ops.aten.unsqueeze.default(unsqueeze_6, -1);  unsqueeze_6 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         add_1: f32[3, 32, 15, 15] = torch.ops.aten.add.Tensor(mul_2, unsqueeze_7);  mul_2 = unsqueeze_7 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convert_element_type_2: bf16[3, 32, 15, 15] = torch.ops.prims.convert_element_type.default(add_1, torch.bfloat16);  add_1 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/xiaobing/pytorch-offical/test/inductor/test_inductor_freezing.py:335, code: return self.relu6(self.conv_bn(x))
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convert_element_type_3: f32[3, 32, 15, 15] = torch.ops.prims.convert_element_type.default(convert_element_type_2, torch.float32);  convert_element_type_2 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         clamp_min: f32[3, 32, 15, 15] = torch.ops.aten.clamp_min.default(convert_element_type_3, 0.0);  convert_element_type_3 = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         clamp_max: f32[3, 32, 15, 15] = torch.ops.aten.clamp_max.default(clamp_min, 6.0);  clamp_min = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convert_element_type_4: bf16[3, 32, 15, 15] = torch.ops.prims.convert_element_type.default(clamp_max, torch.bfloat16);  clamp_max = None
[2023-08-15 20:19:39,705] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return (convert_element_type_4,)

```

the joint graph path will remove ```convert_element_type_2: bf16[3, 32, 15, 15] = torch.ops.prims.convert_element_type.default(add_1, torch.bfloat16)``` see https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/joint_graph.py#L219, so fo relu6 example, it meet the requirement which conv followed by a series of foldable ops, and then a convert_element_dtype,

Reply:
It's still a little fragile at the moment, and I think we should still do something like I mentioned above. If someone were to add the valid cleanup of :

```

def same_dtype(match):
    return match.output_node().args[0].meta["val"].dtype == match.kwargs["dtype"]

@register_graph_pattern(
    CallFunction(
        torch.ops.prims.convert_element_type.default,
        Ignored(),
        KeywordArg("dtype"),
    ),
    pass_dict=patterns,
    extra_check=same_dtype,
)
def unnecessary_dtype_convert(match: Match, **kwargs):
    """Remove unnecessary dtype conversion op, probably left as a result of Conv-Bn folding"""
    graph = match.graph
    node = match.output_node()
    node.replace_all_uses_with(node.args[0])
    graph.erase_node(node)
```

then this test would start failing. we can still land as is though.